Rewrite ivy page, add template typecheck doc (#33152)
angular/angular: aio/content/guide/ivy.md

> # Migrating to Ivy

Reviewer: ```suggestion
# Angular Ivy
```

Ivy is the code name for Angular's [next-generation compilation and rendering pipeline](https://blog.angular.io/a-plan-for-version-8-0-and-ivy-b3318dfc19f7).
With the version 9 release of Angular, the new compiler and runtime instructions is used by default instead of the older compiler and runtime, known as View Engine.

To try Ivy with an existing project prior to the general release of version 9.0.0, you should update to the latest `@next` release of Angular with the following command:

<code-example language="bash">
ng update @angular/core@next @angular/cli@next
</code-example>

To try Ivy in a new project, install the `@angular/cli@next` package globally and create a new application.

<code-example language="bash">
ng new my-app

> </code-example>

Reviewer: These instructions are already part of our "Updating to version 9" guide (https://github.com/angular/angular/blob/master/aio/content/guide/updating-to-version-9.md#updating-cli-apps). I don't think we need them here, since they're not Ivy-specific

<div class="alert is-helpful">

Learn more about the [Compiler](https://www.youtube.com/watch?v=anphffaCZrQ) and [Runtime](https://www.youtube.com/watch?v=S0o-4yc2n-8) in these videos from our team.

</div>

> ## Opting out of Ivy in version 9

Reviewer: Move to bottom

}
</code-example>

## Ivy and libraries

Ivy applications can be built with libraries that were created with the View Engine compiler.
This compatibility is provided by a tool known as the Angular compatibility compiler (`ngcc`).
CLI commands run `ngcc` as a post-install step as needed when performing an Angular build.

> In particular, the step runs for any command that affects your `node_modules` folder.

Reviewer 2: this doesn't belong / make sense here.
```suggestion
```

If you are a library author, you should keep using the View Engine compiler as of version 9.
By having all libraries continue to use View Engine, you will maintain compatibility with default v9 applications that use Ivy, as well as with applications that have opted to continue using View Engine.

See the [Creating Libraries](guide/creating-libraries) guide for more on how to compile or bundle your Angular library.
When you use the tools integrated into the Angular CLI or `ng-packagr`, your library will always be built the right way automatically.

{@a opting-out-of-angular-ivy}
## Opting out of Ivy in version 9

In version 9, Ivy is the default. For compatibility of current workflows during the update process, you can choose to opt out of Ivy and continue using the previous compiler, View Engine.

> To do this, change the `angularCompilerOptions` in your project's TypeScript configuration, most commonly located at `/tsconfig.app.json`.

Reviewer 2: ```suggestion
To do this, change the `angularCompilerOptions` in your project's TypeScript configuration, most commonly located at `tsconfig.app.json` at the root of the workspace.
```

> The following example shows the new `enableIvy` option, which defaults to true in version 9.

Reviewer 2: ```suggestion
The following example shows how to set the `enableIvy` option to `false`, in order to opt-out of Ivy. The value of this flag is set to `true` as of version 9.
```

<code-example language="json" header="tsconfig.app.json">
{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "outDir": "./out-tsc/app",
    "types": []
  },
  "include": [
    "src/**/*.ts"
  ],
  "exclude": [
    "src/test.ts",
    "src/**/*.spec.ts"
  ],
"angularCompilerOptions": {
 "enableIvy": false
 }
}

> </code-example>

Reviewer 2: we should also mention that "Additionally, you might want to revert the change to the `aot` config property in angular.json described in the Ivy and AOT section above."

}
</code-example>

## Ivy and libraries

Ivy applications can be built with libraries that were created with the View Engine compiler.
This compatibility is provided by a tool known as the Angular compatibility compiler (`ngcc`).
CLI commands run `ngcc` as needed, either after npm installation of dependencies or when performing an Angular build.

If you are a library author, you should keep using the View Engine compiler as of version 9.
By having all libraries continue to use View Engine, you will maintain compatibility with default v9 applications that use Ivy, as well as with applications that have opted to continue using View Engine.

See the [Creating Libraries](guide/creating-libraries) guide for more on how to compile or bundle your Angular library.
When you use the tools integrated into the Angular CLI or `ng-packagr`, your library will always be built the right way automatically.

> {@a opting-out-of-angular-ivy}

Reviewer: It feels like we are missing a section that links to the compatibility guide. We should recommend people try to debug before giving them the opt-out instructions.

}
</code-example>

## Ivy and libraries

Ivy applications can be built with libraries that were created with the View Engine compiler.
This compatibility is provided by a tool known as the Angular compatibility compiler (`ngcc`).
CLI commands run `ngcc` as needed, either after npm installation of dependencies or when performing an Angular build.

If you are a library author, you should keep using the View Engine compiler as of version 9.
By having all libraries continue to use View Engine, you will maintain compatibility with default v9 applications that use Ivy, as well as with applications that have opted to continue using View Engine.

See the [Creating Libraries](guide/creating-libraries) guide for more on how to compile or bundle your Angular library.
When you use the tools integrated into the Angular CLI or `ng-packagr`, your library will always be built the right way automatically.

{@a opting-out-of-angular-ivy}
## Opting out of Ivy in version 9

In version 9, Ivy is the default.
For compatibility with current workflows during the update process, you can choose to opt out of Ivy and continue using the previous compiler, View Engine.

> <div class="alert is-helpful>

Reviewer: This is missing a quotation mark, so the alert is messed up

Author: @kara This already landed, so opened new PR - https://github.com/angular/angular/pull/33513

Before disabling Ivy, check out the debugging recommendations in the [Ivy Compatibility Guide](guide/ivy-compatibility#debugging).

</div>

> To opt out of Ivy, change the `angularCompilerOptions` in your project's TypeScript configuration, most commonly located at `/tsconfig.app.json` at the root of the workspace.

Reviewer 2: ```suggestion
To opt out of Ivy, change the `angularCompilerOptions` in your project's TypeScript configuration, most commonly located in `tsconfig.app.json` at the root of the workspace.
```

}
</code-example>

> ## Migrating libraries to Ivy

Reviewer 3: This section does not mention what one should do as a library consumer if `ngcc` does not correctly compile the library. This seems to be a common enough thing at the moment.

Should we point users to some catch-all ngcc issue, or maybe to submit a new issue (possibly using a ngcc related template)?

Author: This no longer applies to this page (although the information you want should probably go somewhere, perhaps in Using Libraries?)

Reviewer 3: I think we want to direct users in how to identify and report ngcc-specific errors, and where to find and apply workarounds when available (I suppose via the ngcc configuration file).

@petebacondarwin do we have information on this topic in the docs? Are there plans to add it?

Reviewer 4: Well there is this https://next.angular.io/guide/updating-to-version-9...
Perhaps we should have a section in there explaining how ngcc might fail?
I don't think a catch-all ngcc issue is a good idea. We have had some of those in the past and they become unfocussed and unusable very quickly.
I would encourage people to search the issues for their error (just like any other bug in Angular).

> ## Migrating libraries to Ivy

Reviewer: ```suggestion
## Ivy and Libraries
```

Ivy applications can be built with libraries that were created with the View Engine compiler.
This compatibility is provided by a tool known as the Angular compatibility compiler (`ngcc`).
CLI commands run `ngcc` as a post-install step as needed when performing an Angular build.
In particular, the step runs for any command that affects your `node_modules` folder.

If you are a library author, you should keep using the View Engine compiler as of version 9.
By having all libraries continue to use View Engine, you will maintain compatibility with default v9 applications that use Ivy, as well as with applications that have opted to continue using View Engine.

See the [Creating Libraries](guide/creating-libraries) guide for more on how to compile or bundle your Angular library.
When you use the tools integrated into the Angular CLI or `ng-packagr`, your library will always be built the right way automatically.

> ## Template type checking in Ivy

Reviewer: Remove this section and instead add nav link under "AOT compiler". I'll add another one in the "updating to version 9" guide when this gets in
